src/Pos/Generator/BlockEvent.hs

```diff
@@ -41,13 +94,388 @@ data BlockRollbackResult
+              Buildable, Generic, Typeable, NFData, Hashable, Random)
+
+-- | A coefficient in the range [0,1]. Pass it to 'weighted' if you ever get
+-- the chance.
+newtype Chance = Chance {getChance :: Rational}
+
+-- | Generate a boolean that may happen to be of true value.
+byChance :: (Monad m, RandomGen g) => Chance -> RandT g m Bool
+byChance (Chance c) = weighted [(False, 1 - c), (True, c)]
+
+data BlockEventGenParams = BlockEventGenParams
+    { _begpSecrets         :: !AllSecrets
+    , _begpBlockCountMax   :: !BlockCount {- ^ the maximum possible amount of
+    blocks in a BlockApply event. Must be 1 or more. Can be violated by 1 in
+    some cases (see Note on reordering corner cases), so if you specify '52'
+    here, some rare events may contain up to '53' blocks. -}
+    , _begpBlockEventCount :: !BlockEventCount
+    , _begpRollbackChance  :: !Chance
+    , _begpFailureChance   :: !Chance
     }
 
-makeLenses ''BlockEventRollback
+makeLenses ''BlockEventGenParams
+
+{- |
+  Return the range of block indices as a half-open interval (closed on the
+  lower bound, open on the upper bound). For example, for block indices
+  @[0, -2, 4, 7]@ the range is a tuple @(-2, 8)@. The reasons for the interval
+  to be half-open:
+    * it's possible to encode an empty range as @(k, k)@
+    * to compute the amount of elements, one can use simple subtraction
+-}
+
+getBlockIndexRange ::
+       [BlockEvent' BlockIndex]
+    -> (BlockIndex, BlockIndex)
+getBlockIndexRange =
+    maybe (0, 0) (\(lower, upper) -> (lower, upper+1)) . minMaxOf (folded.folded)
+
+-- | Generate a random sequence of block events. The final event is either an
+-- expected failure or a rollback of the entire chain to the initial (empty)
+-- state. There's no forking at the moment.
+genBlockEvents ::
+       (MonadBlockGen ctx m, RandomGen g)
+    => BlockEventGenParams
+    -> RandT g m [BlockEvent]
+genBlockEvents begp = do
+    preBlockEvents <- genBlockEvents' begp
+    let
+        (blockIndexStart, blockIndexEnd) =
+            getBlockIndexRange preBlockEvents
+        blockCount = BlockCount $
+            fromIntegral (blockIndexEnd - blockIndexStart)
+    blocks <- lift $ genBlocks $ BlockGenParams
+        { _bgpSecrets    = begp ^. begpSecrets
+        , _bgpBlockCount = blockCount
+        }
+    let
+        toZeroBased :: BlockIndex -> Int
+        toZeroBased i =
+            getBlockIndex i - getBlockIndex blockIndexStart
+        getBlock :: BlockIndex -> BlundDefault
+        getBlock i =
+            -- Indexing is safe here because by construction we have enough
+            -- blocks for the lookup to succeed.
+            getOldestFirst blocks !! toZeroBased i
+    return $ fmap getBlock <$> preBlockEvents
+
+data GenBlockEventState
+    = GbesStartingState
+    | GbesBlockchain (OldestFirst NE BlockIndex)
+    | GbesExpectedFailure
+
+-- | A worker for 'genBlockEvents' that generates event with block indices
```

Comment:
I meant "worker" in a "worker/wrapper" sense here, but I will use a different word.